pandas/core/groupby/groupby.py

```diff
@@ -5628,7 +5631,24 @@ def sample(
         5  black  5
         2   blue  2
         0    red  0
+
+        >>> df.groupby("a").sample(
+        ...   n=1,
+        ...   n_samples = 2,
+        ...   random_state=1
+        ... )
+                a  b
+        4   black  4
+        5   black  5
+        2    blue  2
+        3    blue  3
+        1     red  1
+        0     red  0
```

Comment:
@joelostblom you are right. I need to make that change
Actually I do not think no_samples parameter is required @mroeschke . We can use something like 
<img width="994" alt="image" src="https://github.com/pandas-dev/pandas/assets/25490130/c958ec64-4cce-43df-9fa9-2d8847a94bf1"> @joelostblom 

I think calling the pd.sample() function inside a for loop takes time, but can we not flatten into a sample_size of `sample_size * replicates`, and then reshape the final df like this implementation above @joelostblom

Reply:
the only concern would be that we are using a single sample distribution for one single huge sample, then reshape it

Reply:
For me it would be preferred if the sample number was returned by default. It is easier to drop that column than to create the `sample_no` solution via a separate step using `np.repeat` as you show above (that could would be hard to grok for a novice).

Regarding creating a big flattened array, how would that work with `replace=False`?